zarr/convenience.py

```diff
@@ -348,3 +350,392 @@ def load(store):
+        If provided, will be used to log progress information.
+    **create_kws
+        Passed through to the create_dataset method when copying an array/dataset.
+
+    Examples
+    --------
+    >>> import h5py
+    >>> import zarr
+    >>> import numpy as np
+    >>> source = h5py.File('data/example.h5', mode='w')
+    >>> foo = source.create_group('foo')
+    >>> baz = foo.create_dataset('bar/baz', data=np.arange(100), chunks=(50,))
+    >>> spam = source.create_dataset('spam', data=np.arange(100, 200), chunks=(30,))
+    >>> zarr.tree(source)
+    /
+     ├── foo
+     │   └── bar
+     │       └── baz (100,) int64
+     └── spam (100,) int64
+    >>> dest = zarr.group()
+    >>> import sys
+    >>> zarr.copy(source['foo'], dest, log=sys.stdout)
+    /foo -> /foo
+    /foo/bar -> /foo/bar
+    /foo/bar/baz -> /foo/bar/baz
+    >>> dest.tree()  # N.B., no spam
+    /
+     └── foo
+         └── bar
+             └── baz (100,) int64
+
+    """
+
+    # setup logging
+    with _LogWriter(log) as log:
+        _copy(log, source, dest, name=name, root=True, shallow=shallow,
+              without_attrs=without_attrs, **create_kws)
+
+
+def _copy(log, source, dest, name, root, shallow, without_attrs, **create_kws):
+
+    # are we copying to/from h5py?
+    source_h5py = source.__module__.startswith('h5py.')
+    dest_h5py = dest.__module__.startswith('h5py.')
+
+    # determine name to copy to
+    if name is None:
+        name = source.name.split('/')[-1]
+        if not name:
+            raise TypeError('source has no name, please provide the `name` '
+                            'parameter to indicate a name to copy to')
+
+    if hasattr(source, 'shape'):
+        # copy a dataset/array
+
+        # setup creation keyword arguments
+        kws = create_kws.copy()
+
+        # setup chunks option, preserve by default
+        kws.setdefault('chunks', source.chunks)
+
+        # setup compression options
+        if source_h5py:
+            if dest_h5py:
+                # h5py -> h5py; preserve compression options by default
+                kws.setdefault('compression', source.compression)
+                kws.setdefault('compression_opts', source.compression_opts)
+                kws.setdefault('shuffle', source.shuffle)
+            else:
+                # h5py -> zarr; use zarr default compression options
+                pass
+        else:
+            if dest_h5py:
+                # zarr -> h5py; use some vaguely sensible defaults
+                kws.setdefault('compression', 'gzip')
+                kws.setdefault('compression_opts', 1)
+                kws.setdefault('shuffle', True)
+            else:
+                # zarr -> zarr; preserve compression options by default
+                kws.setdefault('compressor', source.compressor)
+
+        # create new dataset in destination
+        ds = dest.create_dataset(name, shape=source.shape, dtype=source.dtype, **kws)
```

Comment:
Thoughts on raising if `dest` is a Datastet/Array? Admittedly this would happen already, but it might be worth providing a cleaner error than `AttributeError` and a better message.

Reply:
Now raises ValueError if dest is not a group.